// ng-wizard/esm5/lib/core/wizard-step/ng-wizard-step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NgWizardStep } from '../../utils/interfaces';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["*"];
var NgWizardStepComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NgWizardStepComponent, _super);
    function NgWizardStepComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
NgWizardStepComponent.ɵfac = function NgWizardStepComponent_Factory(t) { return ɵNgWizardStepComponent_BaseFactory(t || NgWizardStepComponent); };
NgWizardStepComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgWizardStepComponent, selectors: [["ng-wizard-step"]], features: [ɵngcc0.ɵɵProvidersFeature([
            { provide: NgWizardStep, useExisting: forwardRef(( /**
                     * @return {?}
                     */function () { return NgWizardStepComponent; })) }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "tab-pane", "step-content", 2, "display", "block"]], template: function NgWizardStepComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } }, styles: [""] });
var ɵNgWizardStepComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(NgWizardStepComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardStepComponent, [{
        type: Component,
        args: [{
                selector: 'ng-wizard-step',
                template: "<div class=\"tab-pane step-content\" style=\"display: block\">\n    <ng-content></ng-content>\n</div>",
                providers: [
                    { provide: NgWizardStep, useExisting: forwardRef(( /**
                             * @return {?}
                             */function () { return NgWizardStepComponent; })) }
                ],
                styles: [""]
            }]
    }], null, null); })();
    return NgWizardStepComponent;
}(NgWizardStep));
export { NgWizardStepComponent };
