// ngx-soap/esm2015/lib/ngx-soap.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import * as ɵngcc0 from '@angular/core';
export class NgxSoapModule {
}
NgxSoapModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxSoapModule });
NgxSoapModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxSoapModule_Factory(t) { return new (t || NgxSoapModule)(); }, imports: [[
            HttpClientModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxSoapModule, { imports: function () { return [HttpClientModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxSoapModule, [{
        type: NgModule,
        args: [{
                imports: [
                    HttpClientModule
                ],
                exports: []
            }]
    }], null, null); })();
