// @cmdap/ng-wizard/esm2015/lib/ng-wizard-error/ng-wizard.error.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgWizardErrorType } from './ng-wizard-error-type.enum';
/**
 * @abstract
 */
export class NgWizardError extends Error {
    /**
     * @protected
     * @param {?} type
     * @param {?} message
     */
    constructor(type, message) {
        super(message);
        this.wizardComponentName = '';
        this.wizardPath = '';
        this.stepComponentName = '';
        this.type = type;
    }
}
if (false) {
    /** @type {?} */
    NgWizardError.prototype.type;
    /** @type {?} */
    NgWizardError.prototype.wizardComponentName;
    /** @type {?} */
    NgWizardError.prototype.wizardPath;
    /** @type {?} */
    NgWizardError.prototype.stepComponentName;
}
export class NoWizardRoute extends NgWizardError {
    /**
     * @param {?} wizardComponentName
     */
    constructor(wizardComponentName) {
        super(NgWizardErrorType.NO_WIZARD_ROUTE, `No route configuration for the ${wizardComponentName} found.`);
        this.wizardComponentName = wizardComponentName;
    }
}
if (false) {
    /** @type {?} */
    NoWizardRoute.prototype.wizardComponentName;
}
export class NoChildRoutes extends NgWizardError {
    /**
     * @param {?} wizardComponentName
     * @param {?} wizardPath
     */
    constructor(wizardComponentName, wizardPath) {
        super(NgWizardErrorType.NO_CHILD_ROUTES, `No child routes for the ${wizardComponentName} found.`);
        this.wizardComponentName = wizardComponentName;
        this.wizardPath = wizardPath;
    }
}
if (false) {
    /** @type {?} */
    NoChildRoutes.prototype.wizardComponentName;
    /** @type {?} */
    NoChildRoutes.prototype.wizardPath;
}
export class NoWsInterface extends NgWizardError {
    /**
     * @param {?} stepComponentName
     */
    constructor(stepComponentName) {
        super(NgWizardErrorType.NO_WS_INTERFACE, `The ${stepComponentName} does not extend the NgWizardStep class or implement the NgWizardStepInterface.`);
        this.stepComponentName = stepComponentName;
    }
}
if (false) {
    /** @type {?} */
    NoWsInterface.prototype.stepComponentName;
}
