// ngx-countdown/esm5/countdown.config.js
import * as tslib_1 from "tslib";
// tslint:disable: no-inferrable-types
import { Injectable, Inject, LOCALE_ID } from '@angular/core';
import { formatDate } from '@angular/common';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
var CountdownGlobalConfig = /** @class */ (function () {
    function CountdownGlobalConfig(locale) {
        var _this = this;
        this.locale = locale;
        this.demand = false;
        this.leftTime = 0;
        this.format = 'HH:mm:ss';
        this.timezone = '+0000';
        this.formatDate = function (_a) {
            var date = _a.date, formatStr = _a.formatStr, timezone = _a.timezone;
            return formatDate(new Date(date), formatStr, _this.locale, timezone || _this.timezone || '+0000');
        };
    }
    CountdownGlobalConfig.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    CountdownGlobalConfig.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function CountdownGlobalConfig_Factory() { return new CountdownGlobalConfig(i0.ɵɵinject(i0.LOCALE_ID)); }, token: CountdownGlobalConfig, providedIn: "root" });
    CountdownGlobalConfig = tslib_1.__decorate([ tslib_1.__param(0, Inject(LOCALE_ID)),
        tslib_1.__metadata("design:paramtypes", [String])
    ], CountdownGlobalConfig);
CountdownGlobalConfig.ɵfac = function CountdownGlobalConfig_Factory(t) { return new (t || CountdownGlobalConfig)(ɵngcc0.ɵɵinject(LOCALE_ID)); };
CountdownGlobalConfig.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CountdownGlobalConfig, factory: function (t) { return CountdownGlobalConfig.ɵfac(t); }, providedIn: 'root' });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountdownGlobalConfig, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: String, decorators: [{
                type: Inject,
                args: [LOCALE_ID]
            }] }]; }, null); })();
    return CountdownGlobalConfig;
}());
export { CountdownGlobalConfig };
