// ngx-countdown/esm5/countdown.timer.js
import * as tslib_1 from "tslib";
import { Injectable, NgZone } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var CountdownTimer = /** @class */ (function () {
    function CountdownTimer(ngZone) {
        this.ngZone = ngZone;
        this.fns = [];
        this.commands = [];
        this.ing = false;
    }
    CountdownTimer.prototype.start = function () {
        var _this = this;
        if (this.ing === true)
            return;
        this.ing = true;
        this.nextTime = +new Date();
        this.ngZone.runOutsideAngular(function () {
            _this.process();
        });
    };
    CountdownTimer.prototype.process = function () {
        var _this = this;
        while (this.commands.length) {
            this.commands.shift()();
        }
        var diff = +new Date() - this.nextTime;
        var count = 1 + Math.floor(diff / 100);
        diff = 100 - (diff % 100);
        this.nextTime += 100 * count;
        for (var i = 0, len = this.fns.length; i < len; i += 2) {
            var frequency = this.fns[i + 1];
            // 100/s
            if (0 === frequency) {
                this.fns[i](count);
                // 1000/s
            }
            else {
                // 先把末位至0，再每次加2
                frequency += 2 * count - 1;
                var step = Math.floor(frequency / 20);
                if (step > 0) {
                    this.fns[i](step);
                }
                // 把末位还原成1
                this.fns[i + 1] = (frequency % 20) + 1;
            }
        }
        if (!this.ing)
            return;
        setTimeout(function () { return _this.process(); }, diff);
    };
    CountdownTimer.prototype.add = function (fn, frequency) {
        var _this = this;
        this.commands.push(function () {
            _this.fns.push(fn);
            _this.fns.push(frequency === 1000 ? 1 : 0);
            _this.ing = true;
        });
        return this;
    };
    CountdownTimer.prototype.remove = function (fn) {
        var _this = this;
        this.commands.push(function () {
            var i = _this.fns.indexOf(fn);
            if (i !== -1) {
                _this.fns.splice(i, 2);
            }
            _this.ing = _this.fns.length > 0;
        });
        return this;
    };
    CountdownTimer.ctorParameters = function () { return [
        { type: NgZone }
    ]; };
    CountdownTimer = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [NgZone])
    ], CountdownTimer);
CountdownTimer.ɵfac = function CountdownTimer_Factory(t) { return new (t || CountdownTimer)(ɵngcc0.ɵɵinject(ɵngcc0.NgZone)); };
CountdownTimer.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CountdownTimer, factory: function (t) { return CountdownTimer.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountdownTimer, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.NgZone }]; }, null); })();
    return CountdownTimer;
}());
export { CountdownTimer };
