// @cmdap/ng-wizard/esm5/lib/ng-wizard-navigation/ng-wizard-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgWizardService } from '../ng-wizard.service';
var NgWizardNavigationComponent = /** @class */ (function () {
    function NgWizardNavigationComponent(service) {
        this.service = service;
    }
    /**
     * @return {?}
     */
    NgWizardNavigationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.stepData$ = this.service.getStepDataChangesAsObservable();
        this.service.getCurrentStepDataAsObservable().subscribe((/**
         * @param {?} stepData
         * @return {?}
         */
        function (stepData) { return _this.currentStepData = stepData; }));
        this.wizardOptions = this.service.wizardOptions;
    };
    /**
     * @param {?} stepData
     * @return {?}
     */
    NgWizardNavigationComponent.prototype.goToStep = /**
     * @param {?} stepData
     * @return {?}
     */
    function (stepData) {
        if ((this.currentStepData && this.currentStepData.options && this.currentStepData.options.disableNavigation)
            || stepData.order >= this.currentStepData.order) {
            return;
        }
        this.service.navigateToStep(stepData);
    };
    NgWizardNavigationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-wizard-navigation',
                    template: "<div class=\"ng-wizard-navigation-container\">\r\n  <nav class=\"ng-wizard-navigation\">\r\n    <ul class=\"ng-wizard-navigation-list\">\r\n      <li *ngFor=\"let stepData of stepData$ | async\" class=\"ng-wizard-navigation-list-item\" (click)=\"goToStep(stepData);\">\r\n        <div *ngIf=\"stepData?.order < currentStepData?.order\"\r\n             [ngClass]=\"{\r\n               'ng-wizard-navigation-link': !currentStepData?.options.disableNavigation,\r\n               'ng-wizard-navigation-disabled': currentStepData?.options.disableNavigation\r\n             }\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.previous\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n\r\n        <div\r\n          *ngIf=\"stepData?.order === currentStepData?.order\"\r\n          class=\"ng-wizard-navigation-active\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.current\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n\r\n        <div *ngIf=\"stepData?.order > currentStepData?.order\"\r\n             class=\"ng-wizard-navigation-disabled\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.next\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n      </li>\r\n    </ul>\r\n  </nav>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    NgWizardNavigationComponent.ctorParameters = function () { return [
        { type: NgWizardService }
    ]; };
    return NgWizardNavigationComponent;
}());
export { NgWizardNavigationComponent };
if (false) {
    /** @type {?} */
    NgWizardNavigationComponent.prototype.stepData$;
    /** @type {?} */
    NgWizardNavigationComponent.prototype.currentStepData;
    /** @type {?} */
    NgWizardNavigationComponent.prototype.wizardOptions;
    /**
     * @type {?}
     * @private
     */
    NgWizardNavigationComponent.prototype.service;
}
