// @cmdap/ng-wizard/esm5/lib/ng-wizard-error/ng-wizard-error.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { NgWizardErrorType } from './ng-wizard-error-type.enum';
import { NgWizardError } from './ng-wizard.error';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgWizardErrorComponent_div_0_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 6);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "br");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelement(4, "br");
    ɵngcc0.ɵɵelementStart(5, "pre");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r695 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" No route configuration for the ", ctx_r695.error.wizardComponentName, " found.");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" Add a route for the ", ctx_r695.error.wizardComponentName, " to your AppRoutingModule.");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1("const routes: Routes = [\n    { path: '', component: ", ctx_r695.error.wizardComponentName, " },\n];\n\n@NgModule({\n  imports: [RouterModule.forRoot(routes)],\n  exports: [RouterModule]\n})\nexport class AppRoutingModule { }");
} }
function NgWizardErrorComponent_div_0_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "br");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelement(4, "br");
    ɵngcc0.ɵɵelementStart(5, "pre");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r696 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" No child routes for the ", ctx_r696.error.wizardComponentName, " found.");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" Add at least one child route for the ", ctx_r696.error.wizardComponentName, " to your AppRoutingModule.");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate2("{ path: '", ctx_r696.error.wizardPath, "', component: ", ctx_r696.error.wizardComponentName, ", children: [\n    { path: 'step1', component: Step1Component },\n    { path: '**', redirectTo: 'step1' },\n] },");
} }
function NgWizardErrorComponent_div_0_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "br");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelement(4, "br");
    ɵngcc0.ɵɵelementStart(5, "pre");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(7, "br");
    ɵngcc0.ɵɵtext(8, " or");
    ɵngcc0.ɵɵelement(9, "br");
    ɵngcc0.ɵɵelementStart(10, "pre");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r697 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" The ", ctx_r697.error.stepComponentName, " does not extend the NgWizardStep class or implement the NgWizardStepInterface.");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" Extend the NgWizardStep class in your ", ctx_r697.error.stepComponentName, " or implement the NgWizardStepInterface.");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1("@NgComponent(...)\nexport class ", ctx_r697.error.stepComponentName, " extends NgWizardStep {\n    constructor() {\n        super();\n    }\n}");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1("@NgComponent(...)\nexport class ", ctx_r697.error.stepComponentName, " implements NgWizardStepInterface {\n    wsIsValid() {\n        return true;\n    }\n    wsOnNext() { }\n    wsOnPrevious() { }\n}");
} }
function NgWizardErrorComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵtemplate(2, NgWizardErrorComponent_div_0_span_2_Template, 7, 3, "span", 3);
    ɵngcc0.ɵɵtemplate(3, NgWizardErrorComponent_div_0_span_3_Template, 7, 4, "span", 4);
    ɵngcc0.ɵɵtemplate(4, NgWizardErrorComponent_div_0_span_4_Template, 12, 4, "span", 5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r694 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r694.error.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r694.NgWizardErrorType.NO_WIZARD_ROUTE);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r694.NgWizardErrorType.NO_CHILD_ROUTES);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r694.NgWizardErrorType.NO_WS_INTERFACE);
} }
var NgWizardErrorComponent = /** @class */ (function () {
    function NgWizardErrorComponent() {
        this.NgWizardErrorType = NgWizardErrorType;
    }
    NgWizardErrorComponent.propDecorators = {
        error: [{ type: Input }]
    };
NgWizardErrorComponent.ɵfac = function NgWizardErrorComponent_Factory(t) { return new (t || NgWizardErrorComponent)(); };
NgWizardErrorComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgWizardErrorComponent, selectors: [["ng-wizard-error"]], inputs: { error: "error" }, decls: 1, vars: 1, consts: [["class", "ng-wizard-error", 4, "ngIf"], [1, "ng-wizard-error"], [1, "ng-wizard-error-message", 3, "ngSwitch"], ["class", "no-wizard-route", 4, "ngSwitchCase"], ["class", "no-child-routes", 4, "ngSwitchCase"], ["class", "no-ws-interface", 4, "ngSwitchCase"], [1, "no-wizard-route"], [1, "no-child-routes"], [1, "no-ws-interface"]], template: function NgWizardErrorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, NgWizardErrorComponent_div_0_Template, 5, 4, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.error);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase], styles: [".ng-wizard-error[_ngcontent-%COMP%]{border:1px solid #d71117;border-radius:5px;color:#d71117;padding:10px 10px 0;margin-top:10px}.ng-wizard-error[_ngcontent-%COMP%]   .ng-wizard-error-message[_ngcontent-%COMP%]   pre[_ngcontent-%COMP%]{display:inline-block;border-radius:3px;background-color:rgba(218,215,197,.4);color:#4d4d4d;padding:15px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardErrorComponent, [{
        type: Component,
        args: [{
                selector: 'ng-wizard-error',
                template: "<div *ngIf=\"error\" class=\"ng-wizard-error\">\r\n  <div [ngSwitch]=\"error.type\" class=\"ng-wizard-error-message\">\r\n\r\n    <!-- NO_WIZARD_ROUTE error -->\r\n\r\n    <span *ngSwitchCase=\"NgWizardErrorType.NO_WIZARD_ROUTE\" class=\"no-wizard-route\">\r\n      No route configuration for the {{ error.wizardComponentName }} found.<br/>\r\n      Add a route for the {{ error.wizardComponentName }} to your AppRoutingModule.<br/>\r\n      <pre>\r\nconst routes: Routes = [\r\n    &#123; path: '', component: {{ error.wizardComponentName }} &#125;,\r\n];\r\n\r\n@NgModule(&#123;\r\n  imports: [RouterModule.forRoot(routes)],\r\n  exports: [RouterModule]\r\n&#125;)\r\nexport class AppRoutingModule &#123; &#125;</pre>\r\n    </span>\r\n\r\n    <!-- NO_CHILD_ROUTES error -->\r\n\r\n    <span *ngSwitchCase=\"NgWizardErrorType.NO_CHILD_ROUTES\" class=\"no-child-routes\">\r\n      No child routes for the {{ error.wizardComponentName }} found.<br/>\r\n      Add at least one child route for the {{ error.wizardComponentName }} to your AppRoutingModule.<br/>\r\n      <pre>\r\n&#123; path: '{{ error.wizardPath }}', component: {{ error.wizardComponentName }}, children: [\r\n    &#123; path: 'step1', component: Step1Component &#125;,\r\n    &#123; path: '**', redirectTo: 'step1' &#125;,\r\n] &#125;,</pre>\r\n    </span>\r\n\r\n    <!-- NO_WS_INTERFACE error -->\r\n\r\n    <span *ngSwitchCase=\"NgWizardErrorType.NO_WS_INTERFACE\" class=\"no-ws-interface\">\r\n      The {{ error.stepComponentName}} does not extend the NgWizardStep class or implement the\r\n      NgWizardStepInterface.<br/>\r\n      Extend the NgWizardStep class in your {{ error.stepComponentName}} or implement the\r\n      NgWizardStepInterface.<br/>\r\n      <pre>\r\n@NgComponent(...)\r\nexport class {{ error.stepComponentName }} extends NgWizardStep &#123;\r\n    constructor() &#123;\r\n        super();\r\n    &#125;\r\n&#125;</pre><br/>\r\n      or<br/>\r\n      <pre>\r\n@NgComponent(...)\r\nexport class {{ error.stepComponentName }} implements NgWizardStepInterface &#123;\r\n    wsIsValid() &#123;\r\n        return true;\r\n    &#125;\r\n    wsOnNext() &#123; &#125;\r\n    wsOnPrevious() &#123; &#125;\r\n&#125;</pre>\r\n    </span>\r\n\r\n  </div>\r\n</div>\r\n",
                styles: [".ng-wizard-error{border:1px solid #d71117;border-radius:5px;color:#d71117;padding:10px 10px 0;margin-top:10px}.ng-wizard-error .ng-wizard-error-message pre{display:inline-block;border-radius:3px;background-color:rgba(218,215,197,.4);color:#4d4d4d;padding:15px}"]
            }]
    }], function () { return []; }, { error: [{
            type: Input
        }] }); })();
    return NgWizardErrorComponent;
}());
export { NgWizardErrorComponent };
if (false) {
    /** @type {?} */
    NgWizardErrorComponent.prototype.error;
    /** @type {?} */
    NgWizardErrorComponent.prototype.NgWizardErrorType;
}
