// @cmdap/ng-wizard/esm2015/lib/ng-wizard.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgWizardComponent } from './ng-wizard.component';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { NgWizardErrorComponent } from './ng-wizard-error/ng-wizard-error.component';
import { NgWizardNavigationComponent } from './ng-wizard-navigation/ng-wizard-navigation.component';
import { NgWizardButtonsComponent } from './ng-wizard-buttons/ng-wizard-buttons.component';
import { NgWizardService } from './ng-wizard.service';
import * as ɵngcc0 from '@angular/core';
export class NgWizardModule {
}
NgWizardModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgWizardModule });
NgWizardModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgWizardModule_Factory(t) { return new (t || NgWizardModule)(); }, providers: [NgWizardService], imports: [[CommonModule, RouterModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgWizardModule, { declarations: function () { return [NgWizardComponent,
        NgWizardErrorComponent,
        NgWizardNavigationComponent,
        NgWizardButtonsComponent]; }, imports: function () { return [CommonModule, RouterModule]; }, exports: function () { return [NgWizardComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgWizardComponent,
                    NgWizardErrorComponent,
                    NgWizardNavigationComponent,
                    NgWizardButtonsComponent,
                ],
                imports: [CommonModule, RouterModule],
                providers: [NgWizardService],
                exports: [NgWizardComponent]
            }]
    }], null, null); })();
