// ng-wizard/esm5/lib/utils/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { STEP_STATE, STEP_STATUS } from './enums';
import { Input, HostBinding } from '@angular/core';
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
export function Language() { }
if (false) {
    /** @type {?|undefined} */
    Language.prototype.next;
    /** @type {?|undefined} */
    Language.prototype.previous;
}
/**
 * @record
 */
export function ToolbarSettings() { }
if (false) {
    /** @type {?|undefined} */
    ToolbarSettings.prototype.toolbarPosition;
    /** @type {?|undefined} */
    ToolbarSettings.prototype.toolbarButtonPosition;
    /** @type {?|undefined} */
    ToolbarSettings.prototype.showNextButton;
    /** @type {?|undefined} */
    ToolbarSettings.prototype.showPreviousButton;
    /** @type {?|undefined} */
    ToolbarSettings.prototype.toolbarExtraButtons;
}
/**
 * @record
 */
export function ToolbarButton() { }
if (false) {
    /** @type {?} */
    ToolbarButton.prototype.text;
    /** @type {?} */
    ToolbarButton.prototype.class;
    /** @type {?|undefined} */
    ToolbarButton.prototype.event;
}
/**
 * @record
 */
export function AnchorSettings() { }
if (false) {
    /** @type {?|undefined} */
    AnchorSettings.prototype.anchorClickable;
    /** @type {?|undefined} */
    AnchorSettings.prototype.enableAllAnchors;
    /** @type {?|undefined} */
    AnchorSettings.prototype.markDoneStep;
    /** @type {?|undefined} */
    AnchorSettings.prototype.markAllPreviousStepsAsDone;
    /** @type {?|undefined} */
    AnchorSettings.prototype.removeDoneStepOnNavigateBack;
    /** @type {?|undefined} */
    AnchorSettings.prototype.enableAnchorOnDoneStep;
}
/**
 * @record
 */
export function NgWizardConfig() { }
if (false) {
    /** @type {?|undefined} */
    NgWizardConfig.prototype.selected;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.keyNavigation;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.cycleSteps;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.lang;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.toolbarSettings;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.anchorSettings;
    /** @type {?|undefined} */
    NgWizardConfig.prototype.theme;
}
/**
 * @abstract
 */
var NgWizardStep = /** @class */ (function () {
    function NgWizardStep() {
    }
    Object.defineProperty(NgWizardStep.prototype, "hidden", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status != STEP_STATUS.active;
        },
        enumerable: true,
        configurable: true
    });
    NgWizardStep.propDecorators = {
        title: [{ type: Input }],
        description: [{ type: Input }],
        state: [{ type: Input }],
        hidden: [{ type: HostBinding, args: ['hidden',] }]
    };
NgWizardStep.ɵfac = function NgWizardStep_Factory(t) { return new (t || NgWizardStep)(); };
NgWizardStep.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgWizardStep, hostVars: 1, hostBindings: function NgWizardStep_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("hidden", ctx.hidden);
    } }, inputs: { title: "title", description: "description", state: "state" } });

    return NgWizardStep;
}());
export { NgWizardStep };
if (false) {
    /** @type {?} */
    NgWizardStep.prototype.index;
    /** @type {?} */
    NgWizardStep.prototype.title;
    /** @type {?} */
    NgWizardStep.prototype.description;
    /** @type {?} */
    NgWizardStep.prototype.state;
    /** @type {?} */
    NgWizardStep.prototype.initialState;
    /** @type {?} */
    NgWizardStep.prototype.status;
    /** @type {?} */
    NgWizardStep.prototype.initialStatus;
}
/**
 * @record
 */
export function StepChangedArgs() { }
if (false) {
    /** @type {?} */
    StepChangedArgs.prototype.step;
    /** @type {?} */
    StepChangedArgs.prototype.previousStep;
    /** @type {?} */
    StepChangedArgs.prototype.direction;
    /** @type {?} */
    StepChangedArgs.prototype.position;
}
