// @cmdap/ng-wizard/esm2015/lib/ng-wizard-buttons/ng-wizard-buttons.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgWizardService } from '../ng-wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgWizardButtonsComponent_div_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r634 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 6);
    ɵngcc0.ɵɵlistener("click", function NgWizardButtonsComponent_div_1_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r634); const currentStepData_r628 = ɵngcc0.ɵɵnextContext().ngIf; const ctx_r632 = ɵngcc0.ɵɵnextContext(); return ctx_r632.goToPath(currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.extra.path); });
    ɵngcc0.ɵɵelement(1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const currentStepData_r628 = ɵngcc0.ɵɵnextContext().ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", currentStepData_r628.options.buttons.extra.label, ɵngcc0.ɵɵsanitizeHtml);
} }
function NgWizardButtonsComponent_div_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r637 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 8);
    ɵngcc0.ɵɵlistener("click", function NgWizardButtonsComponent_div_1_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r637); const ctx_r636 = ɵngcc0.ɵɵnextContext(2); return ctx_r636.goToPreviousStep(); });
    ɵngcc0.ɵɵelement(1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const currentStepData_r628 = ɵngcc0.ɵɵnextContext().ngIf;
    const ctx_r630 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", (currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.previous == null ? null : currentStepData_r628.options.buttons.previous.label) || ctx_r630.wizardOptions.buttons.previous.label, ɵngcc0.ɵɵsanitizeHtml);
} }
function NgWizardButtonsComponent_div_1_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r640 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 9);
    ɵngcc0.ɵɵlistener("click", function NgWizardButtonsComponent_div_1_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r640); const ctx_r639 = ɵngcc0.ɵɵnextContext(2); return ctx_r639.goToNextStep(); });
    ɵngcc0.ɵɵelement(1, "span", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const currentStepData_r628 = ɵngcc0.ɵɵnextContext().ngIf;
    const ctx_r631 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", (currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.next == null ? null : currentStepData_r628.options.buttons.next.label) || ctx_r631.wizardOptions.buttons.next.label, ɵngcc0.ɵɵsanitizeHtml);
} }
function NgWizardButtonsComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵtemplate(1, NgWizardButtonsComponent_div_1_button_1_Template, 2, 1, "button", 3);
    ɵngcc0.ɵɵtemplate(2, NgWizardButtonsComponent_div_1_button_2_Template, 2, 1, "button", 4);
    ɵngcc0.ɵɵtemplate(3, NgWizardButtonsComponent_div_1_button_3_Template, 2, 1, "button", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const currentStepData_r628 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.extra);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", currentStepData_r628.previousStep && !(currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.previous == null ? null : currentStepData_r628.options.buttons.previous.hidden));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", currentStepData_r628.nextStep && !(currentStepData_r628.options.buttons == null ? null : currentStepData_r628.options.buttons.next == null ? null : currentStepData_r628.options.buttons.next.hidden));
} }
export class NgWizardButtonsComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.currentStepData$ = this.service.getCurrentStepDataAsObservable();
        this.wizardOptions = this.service.wizardOptions;
    }
    /**
     * @return {?}
     */
    goToNextStep() {
        this.service.navigateToNextStep();
    }
    /**
     * @return {?}
     */
    goToPreviousStep() {
        this.service.navigateToPreviousStep();
    }
    /**
     * @param {?} path
     * @return {?}
     */
    goToPath(path) {
        this.service.navigateToPath(path);
    }
}
NgWizardButtonsComponent.ɵfac = function NgWizardButtonsComponent_Factory(t) { return new (t || NgWizardButtonsComponent)(ɵngcc0.ɵɵdirectiveInject(NgWizardService)); };
NgWizardButtonsComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgWizardButtonsComponent, selectors: [["ng-wizard-buttons"]], decls: 3, vars: 3, consts: [[1, "ng-wizard-buttons-container"], ["class", "ng-wizard-buttons", 4, "ngIf"], [1, "ng-wizard-buttons"], ["class", "ng-wizard-button-extra", 3, "click", 4, "ngIf"], ["class", "ng-wizard-button-previous", 3, "click", 4, "ngIf"], ["class", "ng-wizard-button-next", 3, "click", 4, "ngIf"], [1, "ng-wizard-button-extra", 3, "click"], [1, "ng-wizard-button-label", 3, "innerHTML"], [1, "ng-wizard-button-previous", 3, "click"], [1, "ng-wizard-button-next", 3, "click"]], template: function NgWizardButtonsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, NgWizardButtonsComponent_div_1_Template, 4, 3, "div", 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 1, ctx.currentStepData$));
    } }, directives: [ɵngcc1.NgIf], pipes: [ɵngcc1.AsyncPipe], styles: [""] });
/** @nocollapse */
NgWizardButtonsComponent.ctorParameters = () => [
    { type: NgWizardService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardButtonsComponent, [{
        type: Component,
        args: [{
                selector: 'ng-wizard-buttons',
                template: "<div class=\"ng-wizard-buttons-container\">\r\n  <div *ngIf=\"currentStepData$ | async as currentStepData\" class=\"ng-wizard-buttons\">\r\n    <button *ngIf=\"currentStepData.options.buttons?.extra\"\r\n            (click)=\"goToPath(currentStepData.options.buttons?.extra.path)\"\r\n            class=\"ng-wizard-button-extra\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons.extra.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n    <button *ngIf=\"currentStepData.previousStep && !currentStepData.options.buttons?.previous?.hidden\"\r\n            (click)=\"goToPreviousStep()\"\r\n            class=\"ng-wizard-button-previous\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons?.previous?.label || wizardOptions.buttons.previous.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n    <button *ngIf=\"currentStepData.nextStep && !currentStepData.options.buttons?.next?.hidden\"\r\n            (click)=\"goToNextStep()\"\r\n            class=\"ng-wizard-button-next\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons?.next?.label || wizardOptions.buttons.next.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n  </div>\r\n</div>\r\n",
                styles: [""]
            }]
    }], function () { return [{ type: NgWizardService }]; }, null); })();
if (false) {
    /** @type {?} */
    NgWizardButtonsComponent.prototype.currentStepData$;
    /** @type {?} */
    NgWizardButtonsComponent.prototype.wizardOptions;
    /**
     * @type {?}
     * @private
     */
    NgWizardButtonsComponent.prototype.service;
}
