// @cmdap/ng-wizard/esm2015/lib/ng-wizard-buttons/ng-wizard-buttons.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgWizardService } from '../ng-wizard.service';
export class NgWizardButtonsComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.currentStepData$ = this.service.getCurrentStepDataAsObservable();
        this.wizardOptions = this.service.wizardOptions;
    }
    /**
     * @return {?}
     */
    goToNextStep() {
        this.service.navigateToNextStep();
    }
    /**
     * @return {?}
     */
    goToPreviousStep() {
        this.service.navigateToPreviousStep();
    }
    /**
     * @param {?} path
     * @return {?}
     */
    goToPath(path) {
        this.service.navigateToPath(path);
    }
}
NgWizardButtonsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-wizard-buttons',
                template: "<div class=\"ng-wizard-buttons-container\">\r\n  <div *ngIf=\"currentStepData$ | async as currentStepData\" class=\"ng-wizard-buttons\">\r\n    <button *ngIf=\"currentStepData.options.buttons?.extra\"\r\n            (click)=\"goToPath(currentStepData.options.buttons?.extra.path)\"\r\n            class=\"ng-wizard-button-extra\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons.extra.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n    <button *ngIf=\"currentStepData.previousStep && !currentStepData.options.buttons?.previous?.hidden\"\r\n            (click)=\"goToPreviousStep()\"\r\n            class=\"ng-wizard-button-previous\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons?.previous?.label || wizardOptions.buttons.previous.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n    <button *ngIf=\"currentStepData.nextStep && !currentStepData.options.buttons?.next?.hidden\"\r\n            (click)=\"goToNextStep()\"\r\n            class=\"ng-wizard-button-next\">\r\n      <span\r\n        [innerHTML]=\"currentStepData.options.buttons?.next?.label || wizardOptions.buttons.next.label\"\r\n        class=\"ng-wizard-button-label\"></span>\r\n    </button>\r\n  </div>\r\n</div>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
NgWizardButtonsComponent.ctorParameters = () => [
    { type: NgWizardService }
];
if (false) {
    /** @type {?} */
    NgWizardButtonsComponent.prototype.currentStepData$;
    /** @type {?} */
    NgWizardButtonsComponent.prototype.wizardOptions;
    /**
     * @type {?}
     * @private
     */
    NgWizardButtonsComponent.prototype.service;
}
