// @cmdap/ng-wizard/esm5/lib/ng-wizard.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgWizardService } from './ng-wizard.service';
var NgWizardComponent = /** @class */ (function () {
    function NgWizardComponent(service) {
        this.service = service;
        try {
            this.service.loadWizardRoutes(this.constructor.name);
        }
        catch (error) {
            this.error = error;
        }
    }
    /**
     * @param {?} componentRef
     * @return {?}
     */
    NgWizardComponent.prototype.onActivate = /**
     * @param {?} componentRef
     * @return {?}
     */
    function (componentRef) {
        try {
            this.service.registerActiveComponent(componentRef);
        }
        catch (error) {
            this.error = error;
        }
    };
    NgWizardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-wizard',
                    template: "<div class=\"ng-wizard-container\">\r\n  <ng-wizard-error [error]=\"error\"></ng-wizard-error>\r\n  <ng-wizard-navigation></ng-wizard-navigation>\r\n  <div class=\"ng-wizard-content-container\">\r\n    <router-outlet (activate)=\"onActivate($event)\"></router-outlet>\r\n  </div>\r\n  <ng-wizard-buttons></ng-wizard-buttons>\r\n</div>\r\n"
                }] }
    ];
    /** @nocollapse */
    NgWizardComponent.ctorParameters = function () { return [
        { type: NgWizardService }
    ]; };
    return NgWizardComponent;
}());
export { NgWizardComponent };
if (false) {
    /** @type {?} */
    NgWizardComponent.prototype.error;
    /**
     * @type {?}
     * @private
     */
    NgWizardComponent.prototype.service;
}
