// @cmdap/ng-wizard/esm2015/lib/ng-wizard-navigation/ng-wizard-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NgWizardService } from '../ng-wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = function (a0, a1) { return { "ng-wizard-navigation-link": a0, "ng-wizard-navigation-disabled": a1 }; };
function NgWizardNavigationComponent_li_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelement(1, "span", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const stepData_r618 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r619 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c0, !(ctx_r619.currentStepData == null ? null : ctx_r619.currentStepData.options.disableNavigation), ctx_r619.currentStepData == null ? null : ctx_r619.currentStepData.options.disableNavigation));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", stepData_r618.options.icon || ctx_r619.wizardOptions.navBar.icons.previous, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", stepData_r618.options.title, " ");
} }
function NgWizardNavigationComponent_li_3_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "span", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const stepData_r618 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r620 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", stepData_r618.options.icon || ctx_r620.wizardOptions.navBar.icons.current, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", stepData_r618.options.title, " ");
} }
function NgWizardNavigationComponent_li_3_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelement(1, "span", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const stepData_r618 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r621 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", stepData_r618.options.icon || ctx_r621.wizardOptions.navBar.icons.next, ɵngcc0.ɵɵsanitizeHtml);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", stepData_r618.options.title, " ");
} }
function NgWizardNavigationComponent_li_3_Template(rf, ctx) { if (rf & 1) {
    const _r626 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 4);
    ɵngcc0.ɵɵlistener("click", function NgWizardNavigationComponent_li_3_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r626); const stepData_r618 = ctx.$implicit; const ctx_r625 = ɵngcc0.ɵɵnextContext(); return ctx_r625.goToStep(stepData_r618); });
    ɵngcc0.ɵɵtemplate(1, NgWizardNavigationComponent_li_3_div_1_Template, 4, 6, "div", 5);
    ɵngcc0.ɵɵtemplate(2, NgWizardNavigationComponent_li_3_div_2_Template, 4, 2, "div", 6);
    ɵngcc0.ɵɵtemplate(3, NgWizardNavigationComponent_li_3_div_3_Template, 4, 2, "div", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const stepData_r618 = ctx.$implicit;
    const ctx_r617 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (stepData_r618 == null ? null : stepData_r618.order) < (ctx_r617.currentStepData == null ? null : ctx_r617.currentStepData.order));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (stepData_r618 == null ? null : stepData_r618.order) === (ctx_r617.currentStepData == null ? null : ctx_r617.currentStepData.order));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (stepData_r618 == null ? null : stepData_r618.order) > (ctx_r617.currentStepData == null ? null : ctx_r617.currentStepData.order));
} }
export class NgWizardNavigationComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.stepData$ = this.service.getStepDataChangesAsObservable();
        this.service.getCurrentStepDataAsObservable().subscribe((/**
         * @param {?} stepData
         * @return {?}
         */
        stepData => this.currentStepData = stepData));
        this.wizardOptions = this.service.wizardOptions;
    }
    /**
     * @param {?} stepData
     * @return {?}
     */
    goToStep(stepData) {
        if ((this.currentStepData && this.currentStepData.options && this.currentStepData.options.disableNavigation)
            || stepData.order >= this.currentStepData.order) {
            return;
        }
        this.service.navigateToStep(stepData);
    }
}
NgWizardNavigationComponent.ɵfac = function NgWizardNavigationComponent_Factory(t) { return new (t || NgWizardNavigationComponent)(ɵngcc0.ɵɵdirectiveInject(NgWizardService)); };
NgWizardNavigationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgWizardNavigationComponent, selectors: [["ng-wizard-navigation"]], decls: 5, vars: 3, consts: [[1, "ng-wizard-navigation-container"], [1, "ng-wizard-navigation"], [1, "ng-wizard-navigation-list"], ["class", "ng-wizard-navigation-list-item", 3, "click", 4, "ngFor", "ngForOf"], [1, "ng-wizard-navigation-list-item", 3, "click"], [3, "ngClass", 4, "ngIf"], ["class", "ng-wizard-navigation-active", 4, "ngIf"], ["class", "ng-wizard-navigation-disabled", 4, "ngIf"], [3, "ngClass"], [1, "ng-wizard-navigation-icon", 3, "innerHTML"], [1, "ng-wizard-navigation-label"], [1, "ng-wizard-navigation-active"], [1, "ng-wizard-navigation-disabled"]], template: function NgWizardNavigationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "nav", 1);
        ɵngcc0.ɵɵelementStart(2, "ul", 2);
        ɵngcc0.ɵɵtemplate(3, NgWizardNavigationComponent_li_3_Template, 4, 3, "li", 3);
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(4, 1, ctx.stepData$));
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgClass], pipes: [ɵngcc1.AsyncPipe], styles: [""] });
/** @nocollapse */
NgWizardNavigationComponent.ctorParameters = () => [
    { type: NgWizardService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardNavigationComponent, [{
        type: Component,
        args: [{
                selector: 'ng-wizard-navigation',
                template: "<div class=\"ng-wizard-navigation-container\">\r\n  <nav class=\"ng-wizard-navigation\">\r\n    <ul class=\"ng-wizard-navigation-list\">\r\n      <li *ngFor=\"let stepData of stepData$ | async\" class=\"ng-wizard-navigation-list-item\" (click)=\"goToStep(stepData);\">\r\n        <div *ngIf=\"stepData?.order < currentStepData?.order\"\r\n             [ngClass]=\"{\r\n               'ng-wizard-navigation-link': !currentStepData?.options.disableNavigation,\r\n               'ng-wizard-navigation-disabled': currentStepData?.options.disableNavigation\r\n             }\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.previous\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n\r\n        <div\r\n          *ngIf=\"stepData?.order === currentStepData?.order\"\r\n          class=\"ng-wizard-navigation-active\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.current\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n\r\n        <div *ngIf=\"stepData?.order > currentStepData?.order\"\r\n             class=\"ng-wizard-navigation-disabled\">\r\n          <span [innerHTML]=\"stepData.options.icon || wizardOptions.navBar.icons.next\" class=\"ng-wizard-navigation-icon\"></span>\r\n          <span class=\"ng-wizard-navigation-label\">\r\n            {{ stepData.options.title }}\r\n          </span>\r\n        </div>\r\n      </li>\r\n    </ul>\r\n  </nav>\r\n</div>\r\n",
                styles: [""]
            }]
    }], function () { return [{ type: NgWizardService }]; }, null); })();
if (false) {
    /** @type {?} */
    NgWizardNavigationComponent.prototype.stepData$;
    /** @type {?} */
    NgWizardNavigationComponent.prototype.currentStepData;
    /** @type {?} */
    NgWizardNavigationComponent.prototype.wizardOptions;
    /**
     * @type {?}
     * @private
     */
    NgWizardNavigationComponent.prototype.service;
}
